domain/src/main/java/org/oppia/domain/oppialogger/exceptions/ExceptionsController.kt

```diff
@@ -0,0 +1,199 @@
+
+import androidx.lifecycle.LiveData
+import org.oppia.app.model.ExceptionLog
+import org.oppia.app.model.ExceptionLog.ExceptionType
+import org.oppia.app.model.OppiaExceptionLogs
+import org.oppia.data.persistence.PersistentCacheStore
+import org.oppia.domain.oppialogger.ExceptionLogStorageCacheSize
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.data.DataProviders
+import org.oppia.util.logging.ConsoleLogger
+import org.oppia.util.logging.ExceptionLogger
+import org.oppia.util.networking.NetworkConnectionUtil
+import javax.inject.Inject
+
+private const val EXCEPTIONS_CONTROLLER = "Exceptions Controller"
+
+/** Controller for handling exception logging. */
+class ExceptionsController @Inject constructor(
+  private val exceptionLogger: ExceptionLogger,
+  private val dataProviders: DataProviders,
+  cacheStoreFactory: PersistentCacheStore.Factory,
+  private val consoleLogger: ConsoleLogger,
+  private val networkConnectionUtil: NetworkConnectionUtil,
+  @ExceptionLogStorageCacheSize private val exceptionLogStorageCacheSize: Int
+) {
+  private val exceptionLogStore =
+    cacheStoreFactory.create("exception_logs", OppiaExceptionLogs.getDefaultInstance())
+
+  /**
+   * Logs a non-fatal [exception].
+   * Note that exceptions may not actually be logged depending on the network status of the device.
+   * Older exceptions will be pruned to make room for newer exceptions.
+   *
+   * @param timestampInMillis the time, in milliseconds, when the exception occurred
+   */
+  fun logNonFatalException(exception: Exception, timestampInMillis: Long) {
+    uploadOrCacheExceptionLog(exception, timestampInMillis, ExceptionType.NON_FATAL)
+  }
+
+  /**
+   * Logs a fatal [exception].
+   * Note that exceptions may not actually be logged depending on the network status of the device.
+   * Older exceptions will be pruned to make room for newer exceptions.
+   *
+   * @param timestampInMillis the time, in milliseconds, when the exception occurred
+   */
+  fun logFatalException(exception: Exception, timestampInMillis: Long) {
+    uploadOrCacheExceptionLog(exception, timestampInMillis, ExceptionType.FATAL)
+  }
+
+  /**
+   * Checks network connectivity of the device.
+   *
+   * Saves the [exception] to the [exceptionLogStore] in the absence of it.
+   * Uploads to remote service in the presence of it.
+   */
+  private fun uploadOrCacheExceptionLog(
+    exception: Exception,
+    timestampInMillis: Long,
+    exceptionType: ExceptionType
+  ) {
+    when (networkConnectionUtil.getCurrentConnectionStatus()) {
+      NetworkConnectionUtil.ConnectionStatus.NONE ->
+        cacheExceptionLog(
+          convertExceptionToExceptionLog(
+            exception,
+            timestampInMillis,
+            exceptionType
+          )
+        )
+      else -> exceptionLogger.logException(exception)
+    }
+  }
+
+  /** Returns an [ExceptionLog] from a [throwable]. */
+  private fun convertExceptionToExceptionLog(
```

Comment:
Is it worth making this an extension function as well?

Reply:
We'd still have to pass two arguments but yes it looks better. Done.